ledger/src/blockstore.rs

```diff
@@ -1648,6 +1674,113 @@ impl Blockstore {
         false
     }
 
+    /// Returns true if there is no chaining conflict between
+    /// the `shred` and `merkle_root_meta` of the next or previous
+    /// FEC set, or if shreds from the next or previous set are
+    /// yet to be received.
+    ///
+    /// Otherwise return false and add duplicate proof to
+    /// `duplicate_shreds`.
+    fn check_chained_merkle_root_consistency(
+        &self,
+        just_inserted_shreds: &HashMap<ShredId, Shred>,
+        erasure_set: &ErasureSetId,
+        merkle_root_metas: &HashMap<ErasureSetId, WorkingEntry<MerkleRootMeta>>,
+        shred: &Shred,
+        duplicate_shreds: &mut Vec<PossibleDuplicateShred>,
+    ) -> bool {
+        let (slot, fec_set_index) = erasure_set.store_key();
+
+        let next_erasure_set = ErasureSetId::new(slot, fec_set_index + 1);
+        if let Some(next_merkle_root_meta) =
+            merkle_root_metas.get(&next_erasure_set).map(AsRef::as_ref)
+        {
+            let next_shred_id = ShredId::new(
+                slot,
+                next_merkle_root_meta.first_received_shred_index(),
+                next_merkle_root_meta.first_received_shred_type(),
+            );
+            let next_shred =
+                Self::get_shred_from_just_inserted_or_db(self, just_inserted_shreds, next_shred_id)
+                    .expect("Shred indicated by merkle root meta must exist")
+                    .into_owned();
+            let next_shred = Shred::new_from_serialized_shred(next_shred)
+                .expect("Shred indicated by merkle root meta should deserialize");
+
+            if !self.check_chaining(shred, &next_shred, duplicate_shreds) {
+                return false;
+            }
+        }
+
+        if fec_set_index == 0 {
+            // Although the first fec set chains to the last fec set of the parent block,
+            // if this chain is incorrect we do not which block is the duplicate until votes
+            // are received. We instead delay this check until the block reaches duplicate
+            // confirmation.
+            return true;
+        }
+        let prev_erasure_set = ErasureSetId::new(slot, fec_set_index - 1);
```

Comment:
same comment regarding
```
fec_set_index - 1
```

This one is actually more tricky because we need to know `num_data_shreds` in the previous erasure batch.

Maybe blockstore/rocksdb has some api which gives the immediate next or immediate previous stored key directly; and then we can check if those keys are indeed two consecutive erasure batches.